### PR TITLE
Schedule extender

### PR DESCRIPTION
**Pull request by pzheng1025**

### CODE REVIEW

File: csi/src/main.rs [hunk 1]
```diff
@@ -521,28 +513,48 @@ fn main() -> DatenLordResult<()> {
...
+        RunAsRole::Both => {
+            let controller_server = build_grpc_controller_server(
+                &args.end_point,
+                &args.driver_name,
+                Arc::<MetaData>::clone(&md),
+            )?;
+            let worker_server = build_grpc_worker_server(Arc::<MetaData>::clone(&md))?;
+            let node_server = build_grpc_node_server(&args.end_point, &args.driver_name, md)?;
+            run_grpc_servers(
+                vec![controller_server, node_server, worker_server],
```
Comment (pzheng1025) [inline csi/src/main.rs:540]: use slice

File: csi/src/main.rs [hunk 2]
```diff
@@ -521,28 +513,48 @@ fn main() -> DatenLordResult<()> {
...
+                async_server,
+            );
+        }
+        RunAsRole::SchedulerExtender => {
+            let port = args.worker_port;
+            let scheduler_extender = SchdulerExtender::new(
+                Arc::<MetaData>::clone(&md),
+                SocketAddr::new(IpAddr::V4(Ipv4Addr::LOCALHOST), port),
```
Comment (pzheng1025) [inline csi/src/main.rs:548]: Pass in address

Comment (pzheng1025) [reply csi/src/main.rs:548]: Add new cmd parameter

File: csi/src/meta_data.rs [hunk 3]
```diff
@@ -129,6 +129,7 @@ impl MetaData {
                 }
                 RunAsRole::Controller => md.register_to_etcd(CONTROLLER_PREFIX).await?,
                 RunAsRole::Node => md.register_to_etcd(NODE_PREFIX).await?,
+                RunAsRole::SchedulerExtender => (),
```
Comment (pzheng1025) [inline csi/src/meta_data.rs:132]: Register to etcd too.

File: csi/src/main.rs [hunk 4]
```diff
@@ -310,9 +303,12 @@ struct CliArgs {
     pub run_as: RunAsRole,
     /// Etcd address
     pub etcd_address_vec: Vec<String>,
+    /// Scheduler extender port
+    pub scheduler_extender_port: u16,
 }
 
 /// Parse command line arguments
+#[allow(clippy::too_many_lines)]
```
Comment (pwang7) [inline csi/src/main.rs:311]: remove too_many_line lint allow

**Review by pwang7 — CHANGES_REQUESTED** [review]

File: csi/src/main.rs (hunk 4 above)
Comment (pzheng1025) [reply csi/src/main.rs:311]: This function is to parse command arguments. With the increasing number of args, this function exceeds the limitation. But it makes no sense to split this function.

**Review by rogercloud — APPROVED** [review]

File: csi/src/main.rs (hunk 4 above)
Comment (pzheng1025) [reply csi/src/main.rs:309]: add notes

File: csi/src/scheduler_extender.rs [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+            Nodes: args.Nodes,
+            NodeNames: None,
+            FailedNodes: HashMap::new(),
+            Error: "".to_string(),
+        }
+    }
+
+    /// Prioritize candidate nodes
+    const fn prioritize(_args: &ExtenderArgs) -> Vec<HostPriority> {
+        vec![]
```
Comment (pzheng1025) [inline csi/src/scheduler_extender.rs:178]: add TODO

File: scripts/datenlord.yaml [hunk 6]
```diff
@@ -619,7 +613,135 @@ spec:
...
+      containers:
+        - name: datenlord-scheduler
+          image: k8s.gcr.io/kube-scheduler:v1.19.1
+          imagePullPolicy: IfNotPresent
+          args:
+          - kube-scheduler
+          - --config=/datenlord-scheduler/config.yaml
+          - --address=0.0.0.0
+          - --leader-elect=false
+          - --scheduler-name=datenlord-scheduler
```
Comment (pzheng1025) [inline scripts/datenlord.yaml:698]: comment out this line and add comment.

**Review by pwang7 — APPROVED** [review]

**Pull request merged**